Parser/lexer/lexer.c

```diff
@@ -648,40 +692,34 @@ tok_get_normal_mode(struct tok_state *tok, tokenizer_mode* current_tok, struct t
         /* Process the various legal combinations of b"", r"", u"", and f"". */
         int saw_b = 0, saw_r = 0, saw_u = 0, saw_f = 0, saw_t = 0;
         while (1) {
-            if (!(saw_b || saw_u || saw_f) && (c == 'b' || c == 'B'))
+            if (!saw_b && (c == 'b' || c == 'B')) {
                 saw_b = 1;
+            }
             /* Since this is a backwards compatibility support literal we don't
                want to support it in arbitrary order like byte literals. */
-            else if (!(saw_b || saw_u || saw_r || saw_f || saw_t)
-                     && (c == 'u'|| c == 'U')) {
+            else if (!saw_u && (c == 'u'|| c == 'U')) {
                 saw_u = 1;
             }
             /* ur"" and ru"" are not supported */
-            else if (!(saw_r || saw_u) && (c == 'r' || c == 'R')) {
+            else if (!saw_r && (c == 'r' || c == 'R')) {
                 saw_r = 1;
             }
-            else if (!(saw_f || saw_b || saw_u) && (c == 'f' || c == 'F')) {
+            else if (!saw_f && (c == 'f' || c == 'F')) {
                 saw_f = 1;
             }
-            else if (!(saw_t || saw_u) && (c == 't' || c == 'T')) {
+            else if (!saw_t && (c == 't' || c == 'T')) {
                 saw_t = 1;
             }
             else {
                 break;
             }
             c = tok_nextc(tok);
             if (c == '"' || c == '\'') {
-                if (saw_b && saw_t) {
-                    return MAKE_TOKEN(_PyTokenizer_syntaxerror_known_range(
-                        tok, (int)(tok->start + 1 - tok->line_start),
-                        (int)(tok->cur - tok->line_start),
-                        "can't use 'b' and 't' string prefixes together"));
-                }
-                if (saw_f && saw_t) {
-                    return MAKE_TOKEN(_PyTokenizer_syntaxerror_known_range(
-                        tok, (int)(tok->start + 1 - tok->line_start),
-                        (int)(tok->cur - tok->line_start),
-                        "can't use 'f' and 't' string prefixes together"));
+                // Raise error on incompatible string prefixes:
+                int err = maybe_raise_syntax_error_for_string_prefixes(
+                    tok, saw_b, saw_r, saw_u, saw_f, saw_t);
+                if (err >= 0) {
+                    return MAKE_TOKEN(err);
                 }
```

Comment:
I'd probably be best if `maybe_raise_syntax_error_for_string_prefixes` returned -1 on error and 0 on no error. This way we can do this, which feels more natural:

```suggestion
                // Raise error on incompatible string prefixes:
                int status = maybe_raise_syntax_error_for_string_prefixes(
                    tok, saw_b, saw_r, saw_u, saw_f, saw_t);
                if (status < 0) {
                    return MAKE_TOKEN(ERRORTOKEN);
                }
```

Reply:
I tried this initially, to implement this I would have to do this:

```python
return MAKE_SYNTAX_ERROR("u", "t");
```

->

```python
(void)MAKE_SYNTAX_ERROR("u", "t");
return -1;
```

which is also not that great. So, we either have one problem or another. Which one do we prefer to have? :)

Reply:
I prefer the second one to be honest. There's also the option of:

```
int token;
int status = maybe_raise_syntax_error_for_string_prefixes(
    tok, saw_b, saw_r, saw_u, saw_f, saw_t, &token);
if (status < 0) {
    return MAKE_TOKEN(token);
}
```

And then save the result of calling `MAKE_SYNTAX_ERROR` into `*token`. Isn't this overkill though?

Reply:
Ok, refactored!